File: StudyTracker/StudyTracker/StudySession.swift
import Foundation

/// Representa uma sessão de estudo do aluno
struct StudySession: Identifiable, Codable {
    let id: UUID
    var subject: String
    var startDate: Date
    var endDate: Date?
    var notes: String
    
    /// Duração em segundos
    var duration: TimeInterval {
        let end = endDate ?? Date()
        return end.timeIntervalSince(startDate)
    }
    
    /// Duração formatada como "Xh Ym"
    var formattedDuration: String {
        let totalSeconds = Int(duration)
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        
        if hours > 0 {
            return String(format: "%dh %02dmin", hours, minutes)
        } else if minutes > 0 {
            return String(format: "%dmin %02ds", minutes, seconds)
        } else {
            return String(format: "%ds", seconds)
        }
    }
    
    /// Data formatada para exibição
    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "pt_BR")
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter.string(from: startDate)
    }
    
    var isActive: Bool {
        endDate == nil
    }
    
    init(id: UUID = UUID(), subject: String, startDate: Date = Date(), endDate: Date? = nil, notes: String = "") {
        self.id = id
        self.subject = subject
        self.startDate = startDate
        self.endDate = endDate
        self.notes = notes
    }
}

/// Matérias pré-definidas comuns
enum DefaultSubjects {
    static let all: [String] = [
        "Matemática",
        "Português",
        "Física",
        "Química",
        "Biologia",
        "História",
        "Geografia",
        "Inglês",
        "Redação",
        "Filosofia",
        "Sociologia",
        "Programação",
        "Outro"
    ]
}

File: study_tracker_flutter/lib/models/study_session.dart
import 'dart:convert';

/// Representa uma sessão de estudo do aluno.
class StudySession {
  final String id;
  String subject;
  final DateTime startDate;
  DateTime? endDate;
  String notes;

  StudySession({
    required this.id,
    required this.subject,
    required this.startDate,
    this.endDate,
    this.notes = '',
  });

  /// Duração em segundos.
  int get durationSeconds {
    final end = endDate ?? DateTime.now();
    return end.difference(startDate).inSeconds;
  }

  /// Duração formatada como "Xh Ym" ou "Ym Zs".
  String get formattedDuration {
    final total = durationSeconds;
    final hours = total ~/ 3600;
    final minutes = (total % 3600) ~/ 60;
    final seconds = total % 60;

    if (hours > 0) {
      return '${hours}h ${minutes.toString().padLeft(2, '0')}min';
    } else if (minutes > 0) {
      return '${minutes}min ${seconds.toString().padLeft(2, '0')}s';
    } else {
      return '${seconds}s';
    }
  }

  bool get isActive => endDate == null;

  Map<String, dynamic> toJson() => {
        'id': id,
        'subject': subject,
        'startDate': startDate.toIso8601String(),
        'endDate': endDate?.toIso8601String(),
        'notes': notes,
      };

  factory StudySession.fromJson(Map<String, dynamic> json) => StudySession(
        id: json['id'] as String,
        subject: json['subject'] as String,
        startDate: DateTime.parse(json['startDate'] as String),
        endDate: json['endDate'] != null
            ? DateTime.parse(json['endDate'] as String)
            : null,
        notes: json['notes'] as String? ?? '',
      );

  static String encodeList(List<StudySession> sessions) =>
      jsonEncode(sessions.map((s) => s.toJson()).toList());

  static List<StudySession> decodeList(String jsonStr) {
    final list = jsonDecode(jsonStr) as List<dynamic>;
    return list
        .map((e) => StudySession.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}

/// Matérias pré-definidas comuns.
class DefaultSubjects {
  static const List<String> all = [
    'Matemática',
    'Português',
    'Física',
    'Química',
    'Biologia',
    'História',
    'Geografia',
    'Inglês',
    'Redação',
    'Filosofia',
    'Sociologia',
    'Programação',
    'Outro',
  ];
}
